Wikipedia/Code/UIViewController+WMFDynamicHeightPopoverMessage.m

```diff
@@ -38,15 +42,32 @@ - (void)wmf_presentDynamicHeightPopoverViewControllerForSourceRect:(CGRect)sourc
 - (void)wmf_presentDynamicHeightPopoverViewControllerWithTitle:(NSString *)title
                                                        message:(NSString *)message
                                                          width:(CGFloat)width
+                                                      duration:(NSTimeInterval)duration
                                withPresenterConfigurationBlock:(WMFDynamicHeightPopoverPresentationHandler)presenterConfigurationBlock {
 
-    [self presentViewController:
-              [self wmf_dynamicHeightPopoverViewControllerWithTitle:title
-                                                            message:message
-                                                              width:width
-                                    withPresenterConfigurationBlock:presenterConfigurationBlock]
+    if(self.navigationController.visibleViewController != self){
+        return;
+    }
+
+    UIViewController* popoverVC = [self wmf_dynamicHeightPopoverViewControllerWithTitle:title
+                                                                        message:message
+                                                                          width:width
+                                                withPresenterConfigurationBlock:presenterConfigurationBlock];
+    [self presentViewController:popoverVC
                        animated:NO
-                     completion:nil];
+                     completion:^{
+                         if(duration > 0){
+                             [self performSelector:@selector(dismissPopover:) withObject:popoverVC afterDelay:duration];
+                         }
+                     }];
+}
+
+- (void)dismissPopover:(UIViewController*)popoverVC {
+    // Ensure the popover is still the presented view controller.
+    if(self.presentedViewController == popoverVC){
+        [self dismissViewControllerAnimated:YES
```

Comment:
Probably a good time to start using the completion blocks as well - then you can actually wait for the dismissal before showing the next one.